backend/database-sqlite.js

```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+import { mkdir } from 'node:fs/promises'
+import { join, resolve } from 'node:path'
+import sqlite3 from 'sqlite3'
+
+import { checkKey } from './database.js'
+
+let db: any = null
+let readStatement: any = null
+let writeStatement: any = null
+
+const run = (sql, args) => {
+  return new Promise((resolve, reject) => {
+    db.run(sql, args, (err) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve()
+      }
+    })
+  })
+}
+
+export async function initStorage (options: Object = {}): Promise<void> {
+  const { dirname, filename } = options
+  const dataFolder = resolve(dirname)
+
+  await mkdir(dataFolder, { mode: 0o750, recursive: true })
+
+  await new Promise((resolve, reject) => {
+    if (db) {
+      reject(new Error(`The ${filename} SQLite database is already open.`))
+    }
+    db = new sqlite3.Database(join(dataFolder, filename), (err) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve()
+      }
+    })
+  })
+  await run('CREATE TABLE IF NOT EXISTS Data(key TEXT NOT NULL PRIMARY KEY, value TEXT NOT NULL)')
+  console.log('Connected to the %s SQLite database.', filename)
+  readStatement = db.prepare('SELECT value FROM Data WHERE key = ?')
+  writeStatement = db.prepare('REPLACE INTO Data(key, value) VALUES(?, ?)')
+}
+
+// Useful in test hooks.
+export function clear (): Promise<void> {
+  return run('DELETE FROM Data')
+}
+
+export function exportToJSON (): Promise<Object> {
+  return new Promise((resolve, reject) => {
+    db.all('SELECT key, value FROM Data', [], (err, rows) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(Object.fromEntries(rows.map(row => ([row.key, row.value]))))
+      }
+    })
+  })
+}
+
+export function importFromJSON (json: Object): Promise<void> {
+  return new Promise((resolve, reject) => {
+    db.serialize(() => {
+      db.run('BEGIN')
+      Object.keys(json).forEach(key => writeData(key, json[key]))
+      db.run('COMMIT', [], (err) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      })
+    })
+  })
+}
+
+export function readData (key: string): Promise<Buffer | string | void> {
+  return new Promise((resolve, reject) => {
+    checkKey(key)
+    readStatement.get([key], (err, row) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(row?.value)
```

Comment:
Why is there no corresponding `key.startsWith('blob=')` check here?

Reply:
>  Besides, I think these checks are only needed when writing to the DB (e.g. in 'chelonia/db/set', not when reading from it since the database is assumed to cnotain valid key-value pairs

But it is done in `readData` in `database-fs.js`?

Reply:
Do you mean this line?
```js
    .then(buffer => key.startsWith('blob=') ? buffer : buffer.toString('utf8'))
 ```
 It's not a data validation check, just a test to know whether the value must be decoded to UTF-8 before returning

Reply:
Yeah, so I'm wondering why this `resolve(row?.value)` doesn't do a similar decoding

Reply:
That's because sqlite is able to remember data types, in order to return correctly deserialized values when queried. That's actually quite convenient: if a Node buffer was stored, `row.value` will be a Node buffer, but if a string was passed in, then `row.value` will already be a string. 
The same cannot be said for `fs` functions, Node's `fs.readFile()` won't know by itself whether the data should be decoded, the caller has to know it and specify the right encoding (or none if it wants a buffer)